include/scope.hpp

```diff
@@ -115,6 +105,39 @@ class ScopeStack {
       throw NotInScopeError{""};
     }
   }
+
+  /// @brief Adds the `entry` to the top-most scope of the `kind`.
+  /// @tparam Table The type of the table to add the entry.
+  /// @tparam Entry The type of the entry to add.
+  /// @param kind The kind of the scope to add the entry.
+  /// @param table The class member pointer to the table to add the entry.
+  /// @return The added entry if the `id` of the `entry` isn't already in such
+  /// scope; otherwise, the original entry.
+  template <typename Table, typename Entry>
+  std::shared_ptr<Entry> AddEntry_(std::unique_ptr<Entry> entry, ScopeKind kind,
+                                   std::unique_ptr<Table> Scope::*table);
```

Comment:
The template parameter `Table` can be deduced from `Entry`, as it is actually `TableTemplate<Entry>`. I suggest we unify them. 🤓